File: client/shared/src/commonMain/kotlin/dev/kampr/shared/net/AssetLinks.kt
package dev.kampr.shared.net

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

const val ASSET_LINKS_PATH = "/.well-known/assetlinks.json"

// The relation Credential Manager reads. A file that delegates only `common.handle_all_urls` is an
// app-links file and buys a passkey nothing at all.
private const val LOGIN_CREDS = "delegate_permission/common.get_login_creds"

private val json = Json { ignoreUnknownKeys = true; isLenient = true }

// Why Android refused, when the reason is the one thing the phone can see and the node cannot.
//
// A node names the app it will accept a passkey from, by package and by signing certificate. The
// shipped release APK is named by default, so this is silent for anyone who installed Kampr. A
// debug build carries the machine's own debug keystore and a build from source carries whoever
// built it, and for those the node has to be told — which is impossible to guess at and trivial to
// paste, provided something says which fingerprint to paste.
//
// `null` means the file is fine and the failure was something else. Nothing here is a security
// check: it explains a refusal that has already happened.
fun assetLinkComplaint(document: String?, identity: AppIdentity?): String? {
    if (identity == null) return null
    val statements = document?.let { runCatching { json.parseToJsonElement(it).jsonArray }.getOrNull() }
        ?: return "This node publishes no $ASSET_LINKS_PATH, so Android will not create a passkey " +
            "for it. Add this to its config.toml:\n\n[android]\npackage = \"${identity.packageName}\"\n" +
            "fingerprints = [\"${identity.fingerprint}\"]"
    val targets = statements.filterIsInstance<JsonObject>()
        .filter { statement ->
            (statement["relation"] as? JsonArray)
                ?.any { it.jsonPrimitive.content == LOGIN_CREDS } == true
        }
        .mapNotNull { it["target"] as? JsonObject }
        .filter { it["namespace"]?.jsonPrimitive?.content == "android_app" }
    if (targets.isEmpty()) {
        return "This node's $ASSET_LINKS_PATH delegates nothing to an Android app, so Android " +
            "will not create a passkey for it. Its config.toml needs [android] fingerprints."
    }
    val ours = targets.filter { it["package_name"]?.jsonPrimitive?.content == identity.packageName }
    if (ours.isEmpty()) {
        val named = targets.mapNotNull { it["package_name"]?.jsonPrimitive?.content }.joinToString(", ")
        return "This node delegates passkeys to $named, and this app is ${identity.packageName}."
    }
    val fingerprints = ours.flatMap { target ->
        (target["sha256_cert_fingerprints"] as? JsonArray).orEmpty().map { it.jsonPrimitive.content.uppercase() }
    }
    if (fingerprints.none { it == identity.fingerprint.uppercase() }) {
        return "This node names ${identity.packageName} but not the certificate this build is " +
            "signed with. Add it to [android] fingerprints in its config.toml:\n\n" +
            "\"${identity.fingerprint}\""
    }
    return null
}

// Why the ceremony failed, in the order the phone can actually prove things.
//
// A file that is wrong is the cause the phone can read off the node, so it keeps precedence. What
// is left over are two causes the app cannot see, and both are named because neither can be ruled
// out from here (#288).
//
// The first is the app's own half of the association: Credential Manager checks the agreement in
// both directions, and the app's direction is a manifest entry fixed when its APK was compiled. A
// build reaching this line declares *some* site — the control is hidden otherwise — but not
// necessarily this one, and nothing at runtime can change which.
//
// The second is the one probe #170 measured: the node's file is right and Google, which fetches it
// server-side from the public internet, cannot reach the host. The client does not test that and
// must not — asking Google whether Google can see the operator's node is the app phoning a third
// party about the operator's network. `kampr doctor` answers that half and only that half.
//
// The authenticator's own words are kept either way, so there is something to search for.
fun passkeyRefusal(document: String?, identity: AppIdentity?, host: String, reason: String): String {
    if (identity == null) return reason
    assetLinkComplaint(document, identity)?.let { return it }
    return "This node's own setup is right: it names this app and this build's certificate. Two " +
        "things it cannot see decide the rest. This build lists the sites it may hold passkeys " +
        "for in its own manifest, fixed when it was compiled, and $host has to be one of them. " +
        "And Android never reads the node's file itself — Google's Digital Asset Links service " +
        "fetches https://$host$ASSET_LINKS_PATH over the public internet, so a node that only " +
        "resolves on your own network is one it cannot reach. `kampr doctor` there answers the " +
        "second of those and not the first.\n\nAndroid said: $reason"
}

File: crates/kampr-node/src/assetlinks.rs
use crate::config::Android;
use kampr_auth::android::canonical_fingerprint;

/// The SHA-256 of the certificate `~/.android-keystores/kampr-release.jks` holds — the key every
/// release APK is signed with, and therefore the app identity a phone presents to this node.
///
/// It is a public value: it is derivable from any copy of the APK. It is a default rather than a
/// hard-coded answer because two other builds exist and neither is signed with it — a debug build
/// is signed with the machine's own `~/.android/debug.keystore`, and anyone building Kampr from
/// source signs with a keystore of their own. Both put their fingerprint in `[android]` and this
/// file names their app instead. `assetlinks.rs`'s test reads it back out of the artefact rather
/// than trusting this line.
pub const RELEASE_FINGERPRINT: &str =
    "A0:8A:21:84:46:AA:2B:99:08:5C:67:0B:5A:9B:70:32:5E:05:F9:27:CC:DD:12:17:E7:94:63:13:C7:7F:C6:18";

/// The Digital Asset Links statement list, or `None` when this node delegates to no app.
///
/// Built once, by the caller that builds the router, because an unauthenticated endpoint that does
/// work per request is a way to make a node busy for free.
pub fn document(android: &Android) -> Option<String> {
    // The same certificate copied out of two tools is one certificate, wherever the two copies
    // ended up in the list.
    let mut seen = std::collections::HashSet::new();
    let fingerprints: Vec<String> = android
        .fingerprints
        .iter()
        .filter_map(|f| canonical_fingerprint(f))
        .filter(|f| seen.insert(f.clone()))
        .collect();
    if fingerprints.is_empty() || android.package_name.trim().is_empty() {
        return None;
    }
    Some(
        serde_json::json!([{
            // `common.get_login_creds` and nothing else. `common.handle_all_urls` is the app-link
            // relation, and Kampr claims no URL: an app link names its hosts in the manifest at
            // build time, and every operator's node is at a different one.
            "relation": ["delegate_permission/common.get_login_creds"],
            "target": {
                "namespace": "android_app",
                "package_name": android.package_name.trim(),
                "sha256_cert_fingerprints": fingerprints,
            },
        }])
        .to_string(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_shipped_default_is_a_well_formed_digest() {
        assert_eq!(
            canonical_fingerprint(RELEASE_FINGERPRINT).as_deref(),
            Some(RELEASE_FINGERPRINT)
        );
    }

    #[test]
    fn one_certificate_is_one_entry_wherever_its_copies_landed() {
        let android = Android {
            package_name: "dev.kampr.app".into(),
            fingerprints: vec![
                RELEASE_FINGERPRINT.into(),
                "AB".repeat(32),
                RELEASE_FINGERPRINT.to_lowercase(),
            ],
        };
        let document: serde_json::Value =
            serde_json::from_str(&document(&android).expect("a document")).expect("json");
        let fingerprints = document[0]["target"]["sha256_cert_fingerprints"]
            .as_array()
            .expect("fingerprints");
        assert_eq!(fingerprints.len(), 2, "{fingerprints:?}");
        assert_eq!(fingerprints[0], RELEASE_FINGERPRINT);
    }

    #[test]
    fn a_document_naming_no_usable_certificate_is_no_document() {
        for junk in ["", "not-a-fingerprint", "AA:BB", &"AB".repeat(31)] {
            let android = Android {
                package_name: "dev.kampr.app".into(),
                fingerprints: vec![junk.to_string()],
            };
            assert_eq!(document(&android), None, "{junk} is not a certificate digest");
        }
    }
}
